workflow_scripts/protobuf/build_protobuf_unix.sh

```diff
@@ -13,14 +14,18 @@ if [[ -z "$INSTALL_PROTOBUF_PATH" ]]; then
    export INSTALL_PROTOBUF_PATH=/usr
 fi
 
+if [[ -z "$BUILD_TYPE" ]]; then
+   export BUILD_TYPE=Release
+fi
+
 # Build protobuf from source with -fPIC on Unix-like system
 ORIGINAL_PATH=$(pwd)
 cd ..
-wget https://github.com/protocolbuffers/protobuf/releases/download/v3.16.0/protobuf-cpp-3.16.0.tar.gz
-tar -xvf protobuf-cpp-3.16.0.tar.gz
-cd protobuf-3.16.0
+wget https://github.com/protocolbuffers/protobuf/releases/download/v21.1/protobuf-cpp-3.21.1.tar.gz
+tar -xvf protobuf-cpp-3.21.1.tar.gz
+cd protobuf-3.21.1
 mkdir build_source && cd build_source
-cmake ../cmake -DBUILD_SHARED_LIBS=OFF -DCMAKE_INSTALL_PREFIX=$INSTALL_PROTOBUF_PATH -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_POSITION_INDEPENDENT_CODE=ON -Dprotobuf_BUILD_TESTS=OFF -DCMAKE_BUILD_TYPE=Release
+cmake ../cmake -DBUILD_SHARED_LIBS=OFF -DCMAKE_INSTALL_PREFIX=$INSTALL_PROTOBUF_PATH -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_POSITION_INDEPENDENT_CODE=ON -Dprotobuf_BUILD_TESTS=OFF -DCMAKE_BUILD_TYPE=$BUILD_TYPE
```

Comment:
Should we consider Python apps sharing messages with C++ using a native extension (see https://developers.google.com/protocol-buffers/docs/reference/python-generated#sharing-messages and https://github.com/onnx/onnx/pull/4274#issuecomment-1161262117)?

Reply:
Thank you Xavier for the reminder! Yes it seems that the option `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=cpp` should be added.

Just try to understand more context: so current ONNX does not rely on this option with Protobuf 4.21.0+ (Sharing Messages Between Python and C++), but it would be useful for your PR #4274?

Reply:
Right now, every C++ function exposes with pybind11 receives an onnx graph as a serialized buffer and not the graph itself. (like check_model, shape_inference). I think this serialization / deserialization) has a significant cost and should be avoided.

Reply:
The documentation of protobuf is not really clear about that mechanism. I wonder if this flag should be used when building protobuf for python as well as the version linked with onnx. Default version of protobuf is:

```
>>> import google.protobuf.internal.api_implementation as api
>>> api.Type()
'upb'
```